### PR TITLE
vmm: setup hosts, hostname and resolv.conf files

### PR DESCRIPTION
**Pull request by Burning1020**
If containerd has created these three files, it will append relevant cri mounts to container. If these mounts are defined in Kubernetes Pod Yaml by the end user, they will also exist in container mounts. So anyway, in the above cases, vmm-sandboxer should not append any mounts.

If not, vmm-snadboxer could prepare sandbox files and covert them into oci mounts before sandbox starts. The hostname should be set to either pod hostname or host hostname, the hosts should be set as host, and the DNS should be set either from pod spec or host DNS.

Moreover, it's good to set guest DNS if we want to mount some nfs in guest.

### CODE REVIEW

File: vmm/sandbox/src/container/handler/spec.rs [hunk 1]
```diff
@@ -46,6 +55,7 @@ where
         &self,
         sandbox: &mut KuasarSandbox<T>,
     ) -> containerd_sandbox::error::Result<()> {
+        let shared_path = format!("{}/{}", sandbox.base_dir, SHARED_DIR_SUFFIX);
```
Comment (abel-von) [inline vmm/sandbox/src/container/handler/spec.rs:58]: should we replace it with get_sandbox_shared_path?

File: vmm/task/src/container.rs [hunk 2]
```diff
@@ -195,6 +197,26 @@ impl KuasarFactory {
         Self { sandbox }
     }
 
+    // create_sandbox will do some preparation to provide a livable environment for containers,
+    // such as adding guest hooks, setting kernel paras, preparing sandbox files and namespaces
+    pub fn create_sandbox(&self) -> Result<()> {
```
Comment (abel-von) [inline vmm/task/src/container.rs:202]: The things it do has nothing done with `create_sandbox`, I don't think it is a good idea to define such a function here. you can add a static mount and do it in main function, during system initialization

File: vmm/sandbox/src/container/handler/spec.rs (hunk 1 above)
Comment (Burning1020) [reply vmm/sandbox/src/container/handler/spec.rs:58]: Good

File: vmm/task/src/container.rs (hunk 2 above)
Comment (Burning1020) [reply vmm/task/src/container.rs]: I've split it to `early_init_call()` and `late_init_call()` in the second commit.

File: vmm/task/src/main.rs [hunk 3]
```diff
@@ -166,6 +184,22 @@ async fn main() {
...
+// Do some initialization before everything starts.
+// Such as setting envs, preparing cgroup mounts, setting kernel paras.
+async fn early_init_call() -> Result<()> {
+    // Set environment variables from ENVS vector(ordered).
+    for (k, v) in ENVS.iter() {
+        std::env::set_var(k, v);
+    }
+    // Set kernel parameters from KERNEL_PARAS vector(ordered).
+    for (_path, _v) in KERNEL_PARAS.iter() {
+        // TODO: write kernel paras
```
Comment (abel-von) [inline vmm/task/src/main.rs:196]: kernel params may got from cmdline params, The static params may be compiled into the kernel.

File: vmm/task/src/main.rs [hunk 4]
```diff
@@ -258,15 +292,36 @@ async fn init_vm_rootfs() -> containerd_shim::Result<()> {
...
 }
 
-async fn mount_static_mounts(mounts: Vec<StaticMount>) -> containerd_shim::Result<()> {
+async fn mount_static_mounts(mounts: Vec<StaticMount>) -> Result<()> {
     for m in mounts {
-        tokio::fs::create_dir_all(Path::new(m.dest))
-            .await
-            .map_err(io_error!(e, "failed to create {}: ", m.dest))?;
+        let dest = Path::new(m.dest);
+        // For DNS mounts, dest file already exist.
```
Comment (abel-von) [inline vmm/task/src/main.rs:317]: this is a function of common purpose, has nothing to do with DNS mounts, and I think it is wrong to make decision by if the dest exists. maybe we should assume that `mount_static_mounts` only handle dest of type directory. those with file bind mount, you just call the `mount` directly.

File: vmm/task/src/main.rs (hunk 3 above)
Comment (Burning1020) [reply vmm/task/src/main.rs:196]: I'd like to remove code of setting kernel paras as we dont have any paras for now.

File: vmm/task/src/main.rs (hunk 4 above)
Comment (Burning1020) [reply vmm/task/src/main.rs]: Updated

**Review by abel-von — APPROVED** [review]

**Pull request merged**